File: spring-web-java/src/main/java/spring/web/java/presentation/controller/MemberController.java
package spring.web.java.presentation.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.util.ObjectUtils;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import lombok.RequiredArgsConstructor;
import spring.web.java.application.service.MemberService;
import spring.web.java.presentation.dto.request.MemberSaveRequest;
import spring.web.java.presentation.dto.request.MemberUpdateRequest;
import spring.web.java.presentation.dto.response.MemberResponse;
import spring.web.java.presentation.exception.AtLeastOneRequiredException;

@RestController
@RequestMapping("/members")
@RequiredArgsConstructor
public class MemberController {

	private final MemberService memberService;

	@PostMapping
	@ResponseStatus(HttpStatus.CREATED)
	public MemberResponse saveMember(@RequestBody @Validated MemberSaveRequest memberSaveRequest) {
		List<Long> roleIds = memberSaveRequest.roleIds();
		List<Long> permissionIds = memberSaveRequest.permissionIds();

		if (ObjectUtils.isEmpty(roleIds) && ObjectUtils.isEmpty(permissionIds)) {
			throw new AtLeastOneRequiredException("roleIds", "permissionIds");
		}

		return memberService.saveMember(memberSaveRequest);
	}

	@GetMapping
	@PreAuthorize("isAuthenticated()")
	public MemberResponse findMember() {
		return memberService.findMember();
	}

	@PatchMapping
	@PreAuthorize("isAuthenticated()")
	public MemberResponse updateMember(@RequestBody @Validated MemberUpdateRequest memberUpdateRequest) {
		return memberService.updateMember(memberUpdateRequest);
	}

	@DeleteMapping
	@PreAuthorize("isAuthenticated()")
	@ResponseStatus(HttpStatus.NO_CONTENT)
	public void deleteMember() {
		memberService.deleteMember();
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/presentation/controller/MemberController.kt
package spring.web.kotlin.presentation.controller

import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import spring.web.kotlin.presentation.dto.request.MemberSaveRequest
import spring.web.kotlin.presentation.dto.request.MemberUpdateRequest
import spring.web.kotlin.application.service.MemberService
import spring.web.kotlin.presentation.dto.response.MemberResponse
import spring.web.kotlin.presentation.exception.AtLeastOneRequiredException

@RestController
@RequestMapping("members")
class MemberController(
    private val memberService: MemberService,
) {
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun saveMember(@RequestBody @Validated memberSaveRequest: MemberSaveRequest): MemberResponse {
        if (memberSaveRequest.roleIds.isEmpty() && memberSaveRequest.permissionIds.isEmpty()) {
            throw AtLeastOneRequiredException("roleIds", "permissionIds")
        }

        return memberService.saveMember(memberSaveRequest)
    }

    @GetMapping
    @PreAuthorize("isAuthenticated()")
    fun findMember() = memberService.findMember()

    @PatchMapping
    @PreAuthorize("isAuthenticated()")
    fun updateMember(@RequestBody @Validated memberUpdateRequest: MemberUpdateRequest) =
        memberService.updateMember(memberUpdateRequest)

    @DeleteMapping
    @PreAuthorize("isAuthenticated()")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteMember() {
        memberService.deleteMember()
    }
}
